avro-fastserde/src/main/java/com/linkedin/avro/fastserde/PrimitiveFloatList.java

```diff
@@ -0,0 +1,291 @@
+    }
+  }
+
+  public PrimitiveFloatList(Collection<Float> c) {
+    if (c != null) {
+      elements = new float[c.size()];
+      addAll(c);
+    }
+  }
+
+  /**
+   * Instantiate (or re-use) and populate a {@link PrimitiveFloatList} from a {@link org.apache.avro.io.ResolvingDecoder}.
+   *
+   * N.B.: the caller must ensure the data is of the appropriate type by calling {@link #isFloatArray(Schema)}.
+   */
+  public static Object readPrimitiveFloatArray(Object old, Decoder in) throws IOException {
+    long l = in.readArrayStart();
+    if (l > 0) {
+      PrimitiveFloatList array = (PrimitiveFloatList) newPrimitiveFloatArray(old, (int) l);
+      do {
+        for (long i = 0; i < l; i++) {
+          array.addPrimitive(in.readFloat());
+        }
+        l = in.arrayNext();
+      } while (l > 0);
+      return array;
+    } else {
+      return newPrimitiveFloatArray(old, 0);
+    }
+  }
+
+  /**
+   * @param expected {@link Schema} to inspect
+   * @return true is the {@param expected} SCHEMA is of the right type to decode as a {@link PrimitiveFloatList}
+   *         false otherwise
+   */
+  public static boolean isFloatArray(Schema expected) {
+    return expected != null && Schema.Type.ARRAY.equals(expected.getType()) && FLOAT_SCHEMA.equals(
+        expected.getElementType());
+  }
+
+  private static Object newPrimitiveFloatArray(Object old, int size) {
+    if (old instanceof PrimitiveFloatList) {
+      PrimitiveFloatList oldFloatList = (PrimitiveFloatList) old;
+      if (size <= oldFloatList.getCapacity()) {
+        // reuse the float array directly
+        oldFloatList.clear();
+        return old;
+      } else {
+        oldFloatList.resizeAndClear(size);
+        return oldFloatList;
+      }
+    } else {
+      return new PrimitiveFloatList(size);
+    }
+  }
+
+  @Override
+  public Schema getSchema() {
+    return SCHEMA;
+  }
+
+  @Override
+  public int size() {
+    return size;
+  }
+
+  @Override
+  public void clear() {
+    size = 0;
+  }
+
+  private int getCapacity() {
+    return elements.length;
+  }
+
+  private void resizeAndClear(int newSize) {
+    elements = new float[newSize];
+    clear();
+  }
+
+  @Override
+  public Iterator<Float> iterator() {
+    return new Iterator<Float>() {
+      private int position = 0;
+
+      @Override
+      public boolean hasNext() {
+        return position < size;
+      }
+
+      @Override
+      public Float next() {
+        return elements[position++];
+      }
+
+      @Override
+      public void remove() {
+        throw new UnsupportedOperationException();
+      }
+    };
+  }
+
+  public float getPrimitive(int i) {
+    if (i >= size) {
+      throw new IndexOutOfBoundsException("Index " + i + " out of bounds.");
+    }
+    return elements[i];
+  }
+
+  @Override
+  public Float get(int i) {
+    return getPrimitive(i);
+  }
+
+  /**
+   * Add a primitive float inside the list, without boxing.
+   */
+  private boolean addPrimitive(float o) {
```

Comment:
According to an offline discussion with @radai-rosenblatt, he thinks we don't even need this, and that the simple act of storing references to primitive floats instead of object Floats would allow the JVM to optimize away the boxing/unboxing sequence. In any case, it makes the code extra explicit to keep it as such, and we haven't tested whether it's true that the JVM behaves that way, so we should probably keep it this way for now. Perhaps we can leave a TODO to verify if it's enough to simply store the data as primitives in order to benefit from the GC optimization.

Reply:
Sure.